add condition to wait on nodes to schedule app workloads

soer3n opened the pull request:
**What this PR does / why we need it**:

This pr migitate problems observed in #13933. With these changes helm releases managed by applicaton installation resources will first be deployed to clusters when at least one node object is registered. Therefore we can enable atomic release flag for cilium cni app without having concerns because initial nodes are taking too much time. This is required to limit the retries on failure to not break the cluster etcd which can happen without a limitation on retry.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->
/kind bug

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Helm releases managed by application installation resources are first deployed when there is at least one node registered to the user cluster.
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

soer3n committed:
add condition to wait on nodes to schedule app workloads

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
add node to fake client before testing app install controller

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
add test for new implemented condition to wait on nodes for apps

Signed-off-by: soer3n <srenhenning@googlemail.com>

mohamed-rafraf commented on pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go:
```diff
@@ -149,6 +151,14 @@ func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, appI
+		return fmt.Errorf("failed to check if nodes are available: %w", err)
+	}
+	if !nodesAvailable {
+		return errors.New("waiting for nodes to be joined to be able to schedule workloads")
```
I suggest not throwing an error when we are waiting for nodes to join the control plane (it will be an annoying error in the UI)
using `log.Debug()` could be helpful to throw the message

soer3n replied:
I moved it before the reconcile function with respecting your suggestion, This is needed because we need a short duration for reconciling which could be problematic when we would rely on the specified default reconcile duration which could be very long.

soer3n committed:
return nil to use app reconcile interval when no node is joined

* change error message to debug log message
* return nil instead of an error which leads to reconcile with specified interval for app

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
move node check before the reconcile function

* move node check before reconcile function to set a custom requeue duration
** the duration specified in the resource is by default to high and would cause more problems then it would solve

Signed-off-by: soer3n <srenhenning@googlemail.com>

ahmedwaleedmalik commented on pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go:
```diff
@@ -126,6 +131,15 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+		r.log.Debug("waiting for nodes to be joined to be able to schedule workloads")
+		return reconcile.Result{RequeueAfter: initialRequeueDuration}, nil
+	}
+
```
You can move this code on line `123`, right below the following code segment 

```
	if paused {
		return reconcile.Result{}, nil
	}
```

soer3n replied:
Moved it before the reconcile function.

ahmedwaleedmalik commented on pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go:
```diff
@@ -60,6 +62,9 @@ const (
 	maxRetries = 5
+
+	// initialRequeueDuration is the time interval which is used until a node object to schedule workloads is registered in the cluster.
+	initialRequeueDuration = 15 * time.Second
```
A bit nit-picky, but I think 10 would be the sweet spot, we can be a bit more aggressive in re-checking for node availability here.

soer3n replied:
Changed the duration to 10 instead of 15.

ahmedwaleedmalik commented on pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go:
```diff
@@ -126,6 +131,15 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+		return reconcile.Result{}, fmt.Errorf("failed to check if nodes are available: %w", err)
+	}
+	if !nodesAvailable {
+		r.log.Debug("waiting for nodes to be joined to be able to schedule workloads")
```
waiting for nodes to join the cluster to be able to install applications

soer3n replied:
Changed the message to your suggestion.

soer3n committed:
change initialRequeueDuration to 10s and move new func upper

Signed-off-by: soer3n <srenhenning@googlemail.com>

ahmedwaleedmalik reviewed (approved):
/approve
/milestone KKP 2.27
cc @kubermatic/sig-release-managers 

@soer3n please make sure you coordinate with @cnvergence regarding this PR and https://github.com/kubermatic/kubermatic/pull/14017

cnvergence reviewed (approved):
lgtm thanks
